metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala

```diff
@@ -0,0 +1,211 @@
+package scala.meta.internal.implementation
+
+import org.eclipse.{lsp4j => l}
+
+import scala.meta.internal.metals.CodeLensProvider.LensGoSuperCache
+import scala.meta.internal.metals.DefinitionProvider
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.{
+  ClassSignature,
+  MethodSignature,
+  SymbolInformation,
+  SymbolOccurrence,
+  TextDocument,
+  TypeRef,
+  TypeSignature
+}
+
+class SuperMethodProvider(
+    definitionProvider: DefinitionProvider,
+    findSemanticDbWithPathForSymbol: String => Option[TextDocumentWithPath]
+) {
+
+  def findSuperForMethodOrField(
+      msi: SymbolInformation,
+      documentWithPath: TextDocumentWithPath,
+      symbolRole: SymbolOccurrence.Role,
+      findSymbol: String => Option[SymbolInformation],
+      cache: LensGoSuperCache
+  ): Option[l.Location] = {
+    if (symbolRole.isDefinition && (msi.isMethod || msi.isField)) {
+      findSuperForMethodOrFieldChecked(msi, documentWithPath, cache, findSymbol)
+    } else {
+      None
+    }
+  }
+
+  private def getSuperClasses(
+      si: SymbolInformation,
+      findSymbol: String => Option[SymbolInformation],
+      skip: scala.collection.mutable.Set[SymbolInformation]
+  ): List[(SymbolInformation, Option[TextDocumentWithPath])] = {
+    if (skip.exists(_.symbol == si.symbol)) {
+      List()
+    } else {
+      skip += si
+      si.signature match {
+        case clz: ClassSignature =>
+          val parents = clz.parents
+            .collect { case x: TypeRef => x }
+            .filterNot(si =>
+              Set(
+                "scala/AnyRef#",
+                "scala/Serializable#",
+                "java/io/Serializable#",
+                "scala/AnyVal#"
+              ).contains(si.symbol)
+            )
+            .filterNot(p => skip.exists(_.symbol == p.symbol))
+          val mainParents = parents.headOption
```

Comment:
Why a separate case for the first parent? To linearize we should first go up the first parent and then others.